File: src/clients/vue/src/services/api/messageService.ts
import type Message from '@/models/Message';
import { apiClient } from './client';

export async function getChatMessages(
  chatId: string,
  cursor: number | undefined = undefined,
  limit: number = 50,
): Promise<Message[]> {
  const params = new URLSearchParams();

  if (cursor) {
    params.append('cursor', cursor.toString());
  }
  if (limit) {
    params.append('limit', limit.toString());
  }

  const { data } = await apiClient.get<Message[]>(`/chats/${chatId}/messages?${params}`);
  return data;
}

export async function addMessage(
  chatId: string,
  text: string,
  attachmentIds: string[],
): Promise<void> {
  await apiClient.post<Message[]>(`/chats/${chatId}/messages`, {
    text,
    attachmentIds,
  });
}

export async function editMessage(
  chatId: string,
  messageId: number,
  text: string,
  attachmentIds: string[],
): Promise<void> {
  await apiClient.put<Message[]>(`/chats/${chatId}/messages/${messageId}`, {
    text,
    attachmentIds,
  });
}

export async function deleteMessage(chatId: string, messageId: number): Promise<void> {
  await apiClient.get<Message[]>(`/chats/${chatId}/messages/${messageId}`);
}

File: src/gateway/src/Gateway/Services/MessageService.cs
using QuickChat.Gateway.Extensions;
using QuickChat.Gateway.Models;

namespace QuickChat.Gateway.Services;

public class MessageService(Message.MessageClient client, ICurrentUserProvider currentUserProvider)
    : IMessageService
{
    private readonly Message.MessageClient client = client;
    private readonly ICurrentUserProvider currentUserProvider = currentUserProvider;

    public async Task AddMessage(Guid chatId, UpdateMessageModel model)
    {
        AddMessageRequest request =
            new()
            {
                ChatId = chatId.ToString(),
                UserId = currentUserProvider.GetCurrentUserId().ToString(),
                Text = model.Text
            };
        request.AttachmentIds.AddRange(model.AttachmentIds.Select(id => id.ToString()));
        await client.AddMessageAsync(request);
    }

    public async Task DeleteMessage(Guid chatId, long messageId)
    {
        DeleteMessageRequest request =
            new()
            {
                ChatId = chatId.ToString(),
                ActorId = currentUserProvider.GetCurrentUserId().ToString(),
                Id = messageId
            };
        await client.DeleteMessageAsync(request);
    }

    public async Task EditMessage(Guid chatId, long messageId, UpdateMessageModel model)
    {
        EditMessageRequest request =
            new()
            {
                ChatId = chatId.ToString(),
                Id = messageId,
                ActorId = currentUserProvider.GetCurrentUserId().ToString(),
                Text = model.Text
            };
        request.AttachmentIds.AddRange(model.AttachmentIds.Select(id => id.ToString()));
        await client.EditMessageAsync(request);
    }

    public async Task<IList<MessageModel>> GetChatMessages(Guid chatId, int limit, long? cursor)
    {
        GetChatMessagesRequest request = new() { ChatId = chatId.ToString(), Limit = limit };

        if (cursor != null)
        {
            request.Cursor = cursor.Value;
        }

        GetChatMessagesReply reply = await client.GetChatMessagesAsync(request);
        return [.. reply.Messages.Select(m => m.ToModel())];
    }
}
